lib/optimize/RealContentHashPlugin.js

```diff
@@ -178,10 +178,43 @@ class RealContentHashPlugin {
 						}
 					}
 					if (hashToAssets.size === 0) return;
-					const hashRegExp = new RegExp(
-						Array.from(hashToAssets.keys(), quoteMeta).join("|"),
-						"g"
+					const hashRegExps = Array.from(hashToAssets.keys(), quoteMeta).map(
+						hash => new RegExp(hash, "g")
 					);
+
+					/**
+					 * @param {string} str string to be matched against all hashRegExps
+					 * @returns {string[] | null} matches found
+					 */
+					const hashMatch = str => {
+						/** @type {string[]} */
+						const results = [];
+						for (const hashRegExp of hashRegExps) {
```

Comment:
I think there is a another potential improvement - we can store location of found `hashRegExps` and then avoid using `result.replace` in `hashReplace`, , I think it will be faster, but this need to be checked and we can do it latest and merge this as is, because we alreadyh have a good improvement here, so approved

Reply:
@ryanwilsonperkin can you create a separate issue for this. I think this is a great idea but agree this shouldn't block current PR which is a great performance win in itself.

Reply:
Sure thing, captured here: https://github.com/webpack/webpack/issues/16791